### PR TITLE
Tree age

### PR DESCRIPTION
**Pull request by tomkettle1**
Added tree age functionality

### CODE REVIEW

**Comment by tomkettle1** [comment]
Still need to:
- [ ] Add consistent naming
- [ ] Neaten up code
- [ ] Add back in other tree types

**Comment by tomkettle1** [comment]
(Fyi d9b1f5a was just sorting out a small conflict)

**Review by TimMcGilly — CHANGES_REQUESTED** [review]
Seperaratly we need to decide on style and naming convention

File: Tree.cs [hunk 1]
```diff
@@ -12,27 +12,97 @@ interface ITapable
...
     class Tree
     {
+        //TODO: Make abstract
+        public Tree(int age_days)
+        {
+            Age_Days = age_days;
+        }
+        //AGE:
         protected int _age;
-        protected bool _diseased;
```
[the remark this answers is not in the document]
Comment (tomkettle1) [reply Tree.cs:18]: I was gonna take it out until I actually properly added it, but I can jus do that now

[the remark this answers is not in the document]
Comment (TimMcGilly) [reply Tree.cs:18]: Ok, that makes sense putting in another pull request

File: Tree.cs [hunk 2]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
+
+        public int Age_Days
+        {
+            get { return _age; }
+            protected set
+            {
+                if (value >= 0) { _age = value; }
+                else {
+                    throw new ArgumentOutOfRangeException(String.Format("{0} is a negative age.", value), "value");
+                }
```
Comment (tomkettle1) [inline Tree.cs:41]: @TimMcGilly idk if "value" is the right thing for here, but I think they just had the var which was out of range in the example I looked at, which I guess is value

Comment (TimMcGilly) [reply Tree.cs:41]: @tomkettle1 Arguments should be other way round if using two. https://docs.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception.-ctor?view=net-5.0#System_ArgumentOutOfRangeException__ctor_System_String_System_String_

The "value" is the name of incorrect parameter so I would either change "value" to "Age_Days" or just remove it completely as I don't think it adds much value

**Comment by tomkettle1** [comment]
Ok, fixed @TimMcGilly

File: Tree.cs [hunk 3]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
+        public double Age_Years { get { return _age / (double)365; } }
+
+        protected void Increment_Age(int days)
+        {
+            Age_Days += days;
+        }
+        protected void Increment_Age()
+        {
+            Increment_Age(1);
+        }
```
Comment (TimMcGilly) [inline Tree.cs:53]: Should these be public? Currently no way to increment age from outside class as far as I can see

**Review by TimMcGilly — DISMISSED** [review]

File: Tree.cs (hunk 3 above)
Comment (tomkettle1) [reply Tree.cs:53]: I was gonna add (when we get to it) a daily function, that would use these methods. I don't think theirs any reason to have to manually change ages after creation. The only abnormal bit is the 3x when next to the river but I was gonna build that into the daily function

Comment (tomkettle1) [reply Tree.cs:53]: Ok, cos ur busy & we need to add some more pull requests, imma get Ryan to review this & we can add any changes to another pull

File: Tree.cs [hunk 4]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
     {
+        //TODO: Make abstract
+        public Tree(int age_days)
+        {
+            Age_Days = age_days;
+        }
+        //AGE:
         protected int _age;
-        protected bool _diseased;
+        protected (int yearStart, AgeCatagory ageCatagory)[] _ageRanges; // Array of tuples :D
```
Comment (PotatoHax101) [inline Tree.cs:31]:
```suggestion
        protected (int minAge, AgeCatagory ageCatagory)[] _ageRanges; // Array of tuples :D
```

File: Tree.cs [hunk 5]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
+        {
+            Increment_Age(1);
+        }
+
+        public AgeCatagory AgeCatagory {
+            get
+            {
+                for (int i = _ageRanges.Length - 1; i >= 0; i--)
+                {
+                    if (_age >= _ageRanges[i].yearStart)
```
Comment (PotatoHax101) [inline Tree.cs:60]:
```suggestion
                    if (_age >= _ageRanges[i].minAge)
```

File: Tree.cs [hunk 6]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
+                }
+                return AgeCatagory.Unknown;
+            }
+        }
     }
     class Fir : Tree, ICuttable
     {
+        public Fir (int age_days) : base (age_days)
+        {
+            _ageRanges = new (int yearStart, AgeCatagory ageCatagory)[] {
```
Comment (PotatoHax101) [inline Tree.cs:73]:
```suggestion
            _ageRanges = new (int minAge, AgeCatagory ageCatagory)[] {
```

File: Tree.cs [hunk 7]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
         public bool Cut()
         {
             return true;
         }
     }
     class Spruce : Tree, ICuttable
     {
+        public Spruce(int age_days) : base (age_days)
+        {
+            _ageRanges = new (int yearStart, AgeCatagory ageCatagory)[] {
```
Comment (PotatoHax101) [inline Tree.cs:88]:
```suggestion
            _ageRanges = new (int minAge, AgeCatagory ageCatagory)[] {
```

File: Tree.cs [hunk 8]
```diff
@@ -12,27 +12,99 @@ interface ITapable
...
         public bool Cut()
         {
             return true;
         }
     }
     class Maple : Tree, ICuttable, ITapable
     {
+        public Maple(int age_days) : base(age_days)
+        {
+            _ageRanges = new (int yearStart, AgeCatagory ageCatagory)[] {
```
Comment (PotatoHax101) [inline Tree.cs:103]:
```suggestion
            _ageRanges = new (int minAge, AgeCatagory ageCatagory)[] {
```

**Review by PotatoHax101 — CHANGES_REQUESTED** [review]
Code looks good! :D 
Only peace of feedback is to add code comments to make the code easier to follow.

**Review by PotatoHax101 — APPROVED** [review]

**Pull request merged**